### PR TITLE
Add login item API

### LINKED ISSUE #6202: Detect if application is hidden on launch
* Electron version: 1.2.3
* Operating system: OSX 10.11.3

On OSX there is an option to hide an application that's launched automatically on login ```Users & Groups > Login Items```. Electron app does not respect this option and displays the main renderer window when launched at login. 

Is there a way to detect either
- that the app was launched at login, or
- that the app was hidden at launch

?

### PR DESCRIPTION
**Pull request by kevinsawicki**
Adds ~~3~~ 2 new APIs to `app` on macOS 🍎 that allows apps to add and remove themselves as a login item that will open at login.

### Original API

~~Also adds a `getLoginItemStatus` API that reports back if the app is currently set as a login item and if it was currently opened automatically at login.~~

- ~~`app.setAsLoginItem([openAsHidden])`~~
- ~~`app.removeAsLoginItem()`~~
- ~~`app.getLoginItemStatus()`~~

### Updated API

- `app.getLoginItemSettings()` - returns an object with the app's current settings
- `app.setLoginItemSettings(settings)` - sets the app's login item settings, can be used to both add and remove the app as a login item.

/cc @electron/maintainers any naming feedback on these APIs would be appreciated

Closes #6202

### CODE REVIEW

**Comment by zeke** [comment]
What if, instead of `setAsLoginItem` and `removeAsLoginItem`, there was a single `setLoginItemStatus` that expected an object like the one returned by `getLoginItemStatus`?

**Comment by MarshallOfSound** [comment]
Can we maybe make it more generic to allow using the same syntax for Windows.  It's generally called "Start Up" applications on that side of things so maybe.

`app.setAsStartUpItem([openAsHidden])` 
`app.remoteAsStartUpItem()`  
`app.isStartUpItem()`  

Basically just "login items" vs "start up items"

**Comment by kevinsawicki** [comment]
> What if, instead of setAsLoginItem and removeAsLoginItem, there was a single setLoginItemStatus that expected an object like the one returned by getLoginItemStatus?

Yeah, I initially modeled it after the `setAsDefaultProtocolClient` and `removeAsDefaultProtocolClient` API names. A single setter could work though and be similar to other APIs that clear state when parameters aren't specified.

**Comment by kevinsawicki** [comment]
> Can we maybe make it more generic to allow using the same syntax for Windows.

Yeah, good point, I'll take a look at adding this on Windows via the registry and think about a common name more.

**Comment by kevinsawicki** [comment]
> Basically just "login items" vs "start up items"

I think login items could potentially work for both Windows and Mac.

From the Windows docs, https://msdn.microsoft.com/en-us/library/windows/desktop/aa376977(v=vs.85).aspx

> Run and RunOnce registry keys cause programs to run each time that a user logs on.

And although the user facing verbiage on Windows still mentions startup, login items is a more accurate description of the behavior on both platforms.

File: atom/browser/api/atom_api_app.cc [hunk 1]
```diff
@@ -527,8 +527,12 @@ void App::BuildPrototype(
       .SetMethod("show", base::Bind(&Browser::Show, browser))
       .SetMethod("setUserActivity",
                  base::Bind(&Browser::SetUserActivity, browser))
-      .SetMethod("getCurrentActivityType",
-                 base::Bind(&Browser::GetCurrentActivityType, browser))
```
Comment (zcbenz) [inline atom/browser/api/atom_api_app.cc]: Wrong lines removed.

File: atom/browser/browser.h [hunk 2]
```diff
@@ -134,6 +134,15 @@ class Browser : public WindowListObserver {
 
   // Set docks' icon.
   void DockSetIcon(const gfx::Image& image);
+
+  // Get login item status of app
+  v8::Local<v8::Value> GetLoginItemStatus(mate::Arguments* args);
```
Comment (zcbenz) [inline atom/browser/browser.h]: The `atom::Browser` class should not interactive with V8 directly, APIs added here are also meant to be used by non-V8 parts of code. When type conversions and optional arguments are needed, we usually add a wrapper in `api::App`.

File: docs/api/app.md [hunk 3]
```diff
@@ -598,6 +598,35 @@ Returns the current value displayed in the counter badge.
...
+  should restore the state from the previous session. This indicates that the
+  app should restore the windows that were open the last time the app was
+  closed.
+
+### `app.setAsLoginItem([openAsHidden])` _macOS_
+
+Set the app as a login item. This will cause the app to be opened automatically
+at login.
+
+* `openAsHidden` Boolean - `true` to open the app as hidden. Defaults to
```
Comment (zcbenz) [inline docs/api/app.md]: Parameters should be placed above the description.

**Comment by paulcbetts** [comment]
Super easy to implement on Linux too, symlink your .desktop file into `~/.config/autostart`

File: atom/browser/browser_mac.mm [hunk 4]
```diff
@@ -148,6 +150,28 @@
   return prevent_default;
 }
 
+v8::Local<v8::Value> Browser::GetLoginItemStatus(mate::Arguments* args) {
+  bool hidden = false;
+  mate::Dictionary dict = mate::Dictionary::CreateEmpty(args->isolate());
+  dict.Set("openAtLogin", base::mac::CheckLoginItemStatus(&hidden));
+  dict.Set("openAsHidden", hidden);
+  dict.Set("restoreState", base::mac::WasLaunchedAsLoginItemRestoreState());
+  dict.Set("openedAtLogin", base::mac::WasLaunchedAsLoginOrResumeItem());
```
Comment (bengotow) [inline atom/browser/browser_mac.mm]: I was a little confused by the different tenses here—I wonder if calling the keys `wasOpenedAtLogin` or `didOpenAtLogin` might make it more clear?

Comment (kevinsawicki) [inline atom/browser/browser_mac.mm]: > I wonder if calling the keys wasOpenedAtLogin or didOpenAtLogin might make it more clear?

👍 I really like both suggestions, so how about these:

- `openedAtLogin` -> `wasOpenedAtLogin`
- `openedAsHidden` -> `wasOpenedAsHidden`

**Comment by kevinsawicki** [comment]
Thanks everyone for the feedback, the following updates have been made to the APIs.

- `app.getLoginItemStatus()` renamed to `app.getLoginItemSettings()`, same object returned
- `openedAtLogin` renamed to `wasOpenedAtLogin`
- `openedAsHidden` renamed to `wasOpenedAsHidden`
- `app.setAsLoginItem()` and `app.removeAsLoginItem()` have been combined into a `app.setLoginItemSettings()` API that takes a settings Object with the same keys as `app.getLoginItemSettings()` and can be used to add/remove the app as a login item.

**Comment by zcbenz** [comment]
👍

**Pull request merged**